exchanges/order/timeinforce_test.go

```diff
@@ -0,0 +1,127 @@
+package order
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"github.com/thrasher-corp/gocryptotrader/encoding/json"
+)
+
+func TestIsValid(t *testing.T) {
+	t.Parallel()
+	timeInForceValidityMap := map[TimeInForce]bool{
+		TimeInForce(1):    false,
+		ImmediateOrCancel: true,
+		GoodTillTime:      true,
+		GoodTillCancel:    true,
+		GoodTillDay:       true,
+		FillOrKill:        true,
+		PostOnly:          true,
+		UnknownTIF:        true,
+	}
+	var tif TimeInForce
+	for tif = range timeInForceValidityMap {
+		assert.Equalf(t, timeInForceValidityMap[tif], tif.IsValid(), "got %v, expected %v for %v with id %d", tif.IsValid(), timeInForceValidityMap[tif], tif, tif)
+	}
+}
+
+var timeInForceStringToValueMap = map[string]struct {
+	TIF   TimeInForce
+	Error error
+}{
+	"GoodTillCancel":               {TIF: GoodTillCancel},
+	"GOOD_TILL_CANCELED":           {TIF: GoodTillCancel},
+	"GTT":                          {TIF: GoodTillTime},
+	"GOOD_TIL_TIME":                {TIF: GoodTillTime},
+	"FILLORKILL":                   {TIF: FillOrKill},
+	"POST_ONLY_GOOD_TIL_CANCELLED": {TIF: GoodTillCancel | PostOnly},
+	"immedIate_Or_Cancel":          {TIF: ImmediateOrCancel},
+	"IOC":                          {TIF: ImmediateOrCancel},
+	"immediate_or_cancel":          {TIF: ImmediateOrCancel},
+	"IMMEDIATE_OR_CANCEL":          {TIF: ImmediateOrCancel},
+	"IMMEDIATEORCANCEL":            {TIF: ImmediateOrCancel},
+	"GOOD_TILL_CANCELLED":          {TIF: GoodTillCancel},
+	"good_till_day":                {TIF: GoodTillDay},
+	"GOOD_TILL_DAY":                {TIF: GoodTillDay},
+	"GTD":                          {TIF: GoodTillDay},
+	"GOODtillday":                  {TIF: GoodTillDay},
+	"abcdfeg":                      {TIF: UnknownTIF, Error: ErrInvalidTimeInForce},
+	"PoC":                          {TIF: PostOnly},
+	"PendingORCANCEL":              {TIF: PostOnly},
+	"GTX":                          {TIF: GoodTillCrossing},
+	"GOOD_TILL_CROSSING":           {TIF: GoodTillCrossing},
+	"Good Til crossing":            {TIF: GoodTillCrossing},
+}
+
+func TestStringToTimeInForce(t *testing.T) {
+	t.Parallel()
+	for tk := range timeInForceStringToValueMap {
+		result, err := StringToTimeInForce(tk)
+		assert.ErrorIsf(t, err, timeInForceStringToValueMap[tk].Error, "got %v, expected %v", err, timeInForceStringToValueMap[tk].Error)
+		assert.Equalf(t, result, timeInForceStringToValueMap[tk].TIF, "got %v, expected %v", result, timeInForceStringToValueMap[tk].TIF)
+	}
```

Comment:
🚧
* The comments here are duplicating the work of assert
* We're not mentioning what value caused the error
* We're ranging a map, so we should use the map value
* GloriousCode has convinced me to t.Run in these situations
* Equalf params are the wrong way around
* Ideally don't want to ErrorIs `nil`
* Want to avoid making all fields public by default, just as a practice. This is optional.
```suggestion
	for tk, exp := range timeInForceStringToValueMap {
		t.Run(tk, func(t *testing.T){ 
			t.Parallel()
			result, err := StringToTimeInForce(tk)
			if exp.Error != nil {
				require.ErrorIsf(t, err, exp.error)
			} else {
				require.NoError(t, err)
			}
			assert.Equalf(t, exp.v, result)
		}
	}
```